sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteCountingAsynchronousByteChannel.java

```diff
@@ -0,0 +1,138 @@
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+/**
+ * Count bytes written and read to the target channel.
+ */
+public class ByteCountingAsynchronousByteChannel implements AsynchronousByteChannel {
+
+    private final AsynchronousByteChannel channel;
+    private final ProgressReporter readProgressReporter;
+    private final ProgressReporter writeProgressReporter;
+
+    private static final AtomicLongFieldUpdater<ByteCountingAsynchronousByteChannel> BYTES_WRITTEN_ATOMIC_UPDATER =
+        AtomicLongFieldUpdater.newUpdater(ByteCountingAsynchronousByteChannel.class, "bytesWritten");
+    private volatile long bytesWritten;
+    private static final AtomicLongFieldUpdater<ByteCountingAsynchronousByteChannel> BYTES_READ_ATOMIC_UPDATER =
+        AtomicLongFieldUpdater.newUpdater(ByteCountingAsynchronousByteChannel.class, "bytesRead");
+    private volatile long bytesRead;
+
+    public ByteCountingAsynchronousByteChannel(
+        AsynchronousByteChannel channel,
+        ProgressReporter readProgressReporter,
+        ProgressReporter writeProgressReporter) {
+        this.channel = Objects.requireNonNull(channel, "'channel' must not be null");
+        this.readProgressReporter = readProgressReporter;
+        this.writeProgressReporter = writeProgressReporter;
+    }
+
+    @Override
+    public <A> void read(ByteBuffer dst, A attachment, CompletionHandler<Integer, ? super A> handler) {
+        this.channel.read(dst, attachment,
+            new DelegatingCompletionHandler<A>(handler, BYTES_READ_ATOMIC_UPDATER, readProgressReporter));
+    }
+
+    @Override
+    public Future<Integer> read(ByteBuffer dst) {
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        channel.read(dst, dst,
+            new DelegatingCompletionHandler<>(future, BYTES_READ_ATOMIC_UPDATER, readProgressReporter));
+        return future;
+    }
+
+    @Override
+    public <A> void write(ByteBuffer src, A attachment, CompletionHandler<Integer, ? super A> handler) {
+        this.channel.write(src, attachment,
+            new DelegatingCompletionHandler<A>(handler, BYTES_WRITTEN_ATOMIC_UPDATER, writeProgressReporter));
+    }
+
+    @Override
+    public Future<Integer> write(ByteBuffer src) {
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        channel.write(src, src,
+            new DelegatingCompletionHandler<>(future, BYTES_WRITTEN_ATOMIC_UPDATER, writeProgressReporter));
+        return future;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return channel.isOpen();
+    }
+
+    @Override
+    public void close() throws IOException {
+        channel.close();
+    }
+
+    public long getBytesWritten() {
+        return BYTES_WRITTEN_ATOMIC_UPDATER.get(this);
+    }
+
+    public long getBytesRead() {
+        return BYTES_READ_ATOMIC_UPDATER.get(this);
+    }
+
+    private final class DelegatingCompletionHandler<T> implements CompletionHandler<Integer, T> {
```

Comment:
Does this ever work with anything other than `ByteBuffer` as `T`? If not, can we remove the `T`

Reply:
That won't compile because `AsynchronousByteChannel` has interfaces like ` public <A> void write(ByteBuffer src, A attachment, CompletionHandler<Integer, ? super A> handler)` . The super type contract forces us to do this even if we only pass ByteBuffer in our call sites.